p2p/pex/addrbook.go

```diff
@@ -411,8 +411,17 @@ func (a *addrBook) GetSelectionWithBias(biasTowardsNewAddrs int) []*p2p.NetAddre
 	selectionIndex := 0
 ADDRS_LOOP:
 	for selectionIndex < numAddresses {
+
+		// determine whether to pick from an old bucket.
+		// if there's not enough old addresses to pick from, then we cant pick from old bucket.
 		pickFromOldBucket := int((float64(selectionIndex)/float64(numAddresses))*100) >= biasTowardsNewAddrs
-		pickFromOldBucket = (pickFromOldBucket && a.nOld > 0) || a.nNew == 0
+		pickFromOldBucket = (pickFromOldBucket && a.nOld > 0 && len(oldBucketToAddrsMap) < a.nOld) || a.nNew == 0
```

Comment:
Maybe that is overkill but can we used named vars for the conditions here?

sth along the lines of:
```go
isBiasedTowordsOld := int((float64(selectionIndex)/float64(numAddresses))*100) >= biasTowardsNewAddrs
isNoNewAddrs := a.nNew == 0
isOldAddrs := a.nOld > 0
notEnoughOld := len(oldBucketToAddrsMap) < a.nOld

pickFromOldBucket := isBiasedTowordsOld && isOldAddrs && notEnoughOld || isNoNewAddrs
```
(with  better naming)

It would clarify the logic here, makes the code speak for itself etc.

Reply:
```suggestion
		pickFromOldBucket = (pickFromOldBucket &&  a.nOld > len(oldBucketToAddrsMap)) || a.nNew == 0
```

q: what will happen if `a.nNew == 0` -> pickFromOldBucket=true, but we don't have enough items (`a.nOld > len(oldBucketToAddrsMap)` false)?

Reply:
I think it should be 

`pickFromOldBucket = (pickFromOldBucket || a.nNew == 0) && a.nOld > len(oldBucketToAddrsMap)`

Reply:
Note: tests pass for the above code too!

Reply:
I tried to understand this code today for the first time so please bear with me a bit.

"return early" - means there will be less than numAddresses returned, might break this APIs "contract".

Then, if all good addresses (i.e. a.nNew == 0) and pickFromOldBucket is false (because of bias, e.g.first iteration), algo will return right away with empty selection (to be precise, all map entries will point to nil addresses)

I think we should also add tests for all good, all bad, and different sizes so we hit the different cases for numAddresses (bookSize, minGetSelection, bookSize*getSelectionPercent/100). And check that the selection contains in fact the promised number.

Reply:
> "return early" - means there will be less than numAddresses returned, might break this APIs "contract".

GetSelectionWithBias() can return an empty collection. We don't state that it's always returns N items.

Reply:
The problem is we don't state either way. Before this fix, if I have a book of 10 old/ good addresses, and 0 new/ bad, I will get back the 10 old.
With the fix I get back 0. So fixing the infinite loop changed the behavior.

Reply:
Ok, there is this comment:
// GetSelectionWithBias implements AddrBook.
// It randomly selects some addresses (old & new). Suitable for peer-exchange protocols.
**// Must never return a nil address.**

With the example I gave above we return a list of nil addresses.

Reply:
GetSelectionWithBias() returns []*p2p.NetAddress, i.e. a list of pointers to addresses.
- The function determines the number of addresses to return based on the booksize. For small books it will be:
 numAddresses  = a.nNew + a.nOld. 
- Then it allocates that many entries in the list:
selection := make([]*p2p.NetAddress, numAddresses).
I believe the comment: "Must never return a nil address." means that after it runs, all these pointers must point to non-nil addresses.
- Then it starts the loop and gives preference to new addresses if there are any and randomly selects new addresses. At some point it will switch to old bucket and start picking old addresses.

The problem with the latest code in bucky/3230-fix-addrbook-loop is that now the selection only picks new addresses and it will never pick old addresses.

I believe the first commit from Ethan works fine.

Reply:
correction: "it will never pick old addresses." in some cases...in other cases it will do. 
It all depends on the relation between biasToSelectNewPeers and the a.nNew vs a.nOld values.
I think that if there are not enough new addresses (< ~ biasToSelectNewPeers % of total), the old addresses will not be picked.
I don't know, maybe this was the intent all along but the old code did not do that.

I think we need to specify the desired behavior clearly. And then write some code that can convince the reader of correctness via inspection also and not only exhaustive testing.

Reply:
Let' say we decided that we want to pick r (aka numAddresses) addresses.
I think the original intent was:
- pick i new addresses with i=min(n, k), n total number of new addresses, k number based on bias calculation (percentage of r that should be new addresses in the beginning of selection)
- pick j old addresses with j=min(m, r-i)
- i+j < r, pick r-i-j new addresses
(all picks are random)

So we end up with:
```
indices:      0...i-1   i...i+j-1    i+j...r
addresses:    N0..Ni-1  O0..Oj-1     Ni...
```

I will write some testcases for this tomorrow to check if it's true :)

Reply:
> I believe the first commit from Ethan works fine.

ok, reverted my commit

Reply:
With the code from develop:
---------------------------
The algorithm tries to find exactly:

- k new addresses with k ~ bias/100 * r, AND
- r-i old addresses 

When it works, it can only generate:
```
indices:      0...k-1   k...r
addresses:    N0..Nk-1  O0..Or-i-1
```
This is the sequence where we have exactly bias% new addresses and the rest are old.

If the book does NOT have minimum k new addresses AND minimum r-i old addresses, the algo will enter the infite loop.

The exceptions are
- if there are 0 new addresses:
```
indices:      0...r
addresses:    O0..Or-1
```
- if there are 0 old addresses:
```
indices:      0...r
addresses:    N0..Nr-1
```

To give you an idea:
For books of size 100, the computed selection size r is 32. 
For a bias of 30%, the algo needs 10 new addresses and 22 old addresses.
So it will enter infinite loop for any case where number of old addresses m is in {1..21, 91..99}
(32 out of 100 possible combos will enter the infinte loop) 

With the last few fixes:
------------------------
The number of hogs decreased but was not completely eliminated. And now they were also not always reproducible.
The reason is the check involving len(newBucketToAddrsMap) or len(oldBucketToAddrsMap) is not correct, as many addresses are sometimes mapped to the same bucket, so in these cases len(..BucketToAddrsMap returns less than the number of addresses that were added, causing the infinite loop again.


I just committed a fix that maintains separate counters for old and new. I added a test that can take multiple booksizes and run this over all possible combinations. 
The tests need more work, some of the old tests are failing. I am looking at those now.

Reply:
Fixed mistake in bias to old computation ( changed to float division instead of int)